Added cert-controller for easy webhook cert generation

p-strusiewiczsurmacki-mobica opened the pull request:
This PR implements #318 

Certificates can be now generated and injected automatically using [open-policy-agent/cert-controller](https://github.com/open-policy-agent/cert-controller).

chez-shanpu commented on v2/cmd/coil-egress-controller/sub/run.go:
```diff
@@ -78,9 +80,46 @@ func subMain() error {
+		close(mgrErr)
+	}()
+
+	if err := utils.WaitForExit(setupErr, mgrErr, cancel); err != nil {
+		return err
+	}
```
I think the manager always start after the setup is complete, so it seems to me that we don't have to catch `setupErr` and  `mgrErr` at the same time with `utils.WaitForExit()`.

p-strusiewiczsurmacki-mobica replied:
You're right, but it's a bit different with `cert-controller`. It is added and then started by the manager, so `mgr.Start()` has to be called and while manager is already running we need to wait for the certs to be generated so we can add the webhook to the manager.

But I've moved the blocking call right before webhook registration so other things are not blocked.

chez-shanpu replied:
> ... while manager is already running we need to wait for the certs to be generated so we can add the webhook to the manager.

Thanks to your comment, I understand why we have to catch both channels😃
`WaitForExit()` seems to be reasonable.

chez-shanpu commented on v2/cmd/coil-egress-controller/sub/run.go:
```diff
@@ -78,9 +80,46 @@ func subMain() error {
+
+func setupManager(mgr ctrl.Manager, setupFinished chan struct{}) error {
+	// wait for certificates to be configured
+	<-setupFinished
```
I wonder should we block `setupManager()` until certificates are configured.
I think it's enough to use `setupFinished` just before `mgr.Start()`.

p-strusiewiczsurmacki-mobica replied:
As I've stated above (please see the other comment), we had to block at least the webhook registration.

chez-shanpu reviewed (commented):
Thank you for the PR!

I commented to several lines. Please check :)

p-strusiewiczsurmacki-mobica commented on v2/pkg/constants/constants.go:
```diff
@@ -78,6 +78,9 @@ const (
 	DefaultEnableEgress           = true
 	DefaultAddressBlockGCInterval = 5 * time.Minute
+
+	DefaultDisableCertRotation         = false
+	DefailtDisableRestartOnCertRefresh = true
```
Would that not contradict the change above?

```
Coil will generate certificates upon start, but you can also pre-generate those if you want with `make certs` command under `v2/` directory.
```

chez-shanpu replied:
@p-strusiewiczsurmacki-mobica I agree. We should change some of the sentences like you pointed out.

p-strusiewiczsurmacki-mobica replied:
Done :)

p-strusiewiczsurmacki-mobica commented on v2/pkg/utils/utils.go:
```diff
@@ -0,0 +1,70 @@
+package utils
```
Moved that to `pkg/cert`.

p-strusiewiczsurmacki-mobica commented on v2/e2e/suite_test.go:
```diff
@@ -43,6 +44,28 @@ func kubectlSafe(input []byte, args ...string) []byte {
 	return stdout
 }
 
+func kubectlSafeTimeout(input []byte, args ...string) []byte {
```
Done :)

chez-shanpu reviewed (commented):
I left some comments.
And AFAIU, cert-controller isn't tested in e2e on github workflows.
Could you add tests if it is not?

p-strusiewiczsurmacki-mobica commented on v2/config/rbac/coil-egress-controller_role.yaml:
```diff
@@ -24,6 +33,16 @@ rules:
+  - get
+  - list
+  - update
+  - watch
```
Done :)

p-strusiewiczsurmacki-mobica commented on v2/config/rbac/coil-ipam-controller_role.yaml:
```diff
@@ -11,6 +11,25 @@ rules:
+  - get
+  - list
+  - update
+  - watch
```
Done :)

p-strusiewiczsurmacki-mobica commented on v2/config/pod/coil-ipam-controller.yaml:
```diff
@@ -46,6 +46,8 @@ spec:
         args:
           - --zap-stacktrace-level=panic
+          # [CERTS] Following line should be uncommented if automatic cert generation is used.
+          # - --enable-cert-rotation=true
```
I've moved that to separate patch files.

p-strusiewiczsurmacki-mobica commented:
>And AFAIU, cert-controller isn't tested in e2e on github workflows. Could you add tests if it is not?

I've added a `make enable-certs-generation` target that configures the kustomization files and added the tests to the CI.

chez-shanpu commented on v2/config/rbac/coil-egress-controller_role.yaml:
```diff
@@ -12,6 +12,15 @@ rules:
+  - get
+  - list
+  - update
+  - watch
```
How about making a new role resource file for a cert-controller with controller-gen? And apply it (and its role-binding) when `make enable-certs-generation` is executed by using Kustomize's `resource`.
I think this way, we can check with git diff.

p-strusiewiczsurmacki-mobica replied:
Done :)

chez-shanpu reviewed (commented):
I replied to your comment and left a few minor comment :)

chez-shanpu reviewed (commented):
I left a small comment :)

terassyi commented on docs/setup.md:
```diff
@@ -19,17 +21,23 @@ You can tweak optional parameters by editing [`kustomization.yaml`](../v2/kustom
 
-Coil runs an admission webhook server, and it needs a self-signed certificate.
-Run `make certs` under `v2/` directory to generate the certificate.
+Coil runs an admission webhook server, and it needs a self-signed certificate. You can either generate certificates manually or have Coil create them when it starts up.
```
```suggestion
Coil runs admission webhook servers, and each one needs a self-signed certificate. You can either generate certificates manually or have Coil create them when it starts up.
```

p-strusiewiczsurmacki-mobica replied:
Done :)

p-strusiewiczsurmacki-mobica commented on v2/config/rbac/kustomization.yaml:
```diff
@@ -4,13 +4,19 @@ resources:
+
 # [IPAM] Following files should be uncommented to enable IPAM features.
 - coil-ipam-controller_role.yaml
+# - coil-ipam-controller-certs_role.yaml
```
Done :)

chez-shanpu commented on docs/setup.md:
```diff
@@ -8,7 +8,9 @@ The YAML manifests of Coil can be generated using [kustomize](https://kubernetes
-- [Generate TLS certificate](#generate-tls-certificate)
+- [TLS certificates](#tls-certificates)
+  - [Generate certificates manually](#generate-certificates-manually)
+  - [Enable automatic certs generation](#enable-automatic-certs-generation)
```
```suggestion
  - [Enable automatic certs rotation](#enable-automatic-certs-rotation)
```

I think "generation" may be misleading (Coil only updates certificates).

chez-shanpu commented on v2/Makefile:
```diff
@@ -259,3 +291,21 @@ staticcheck:
+	@sed -i -E "{s/(^-.*webhook_manifests_patch.*)/# \1/}" ${$@_FILE}
+	@sed -i -E "{s/(^ +files.*)/# \1/g}" ${$@_FILE}
+	@sed -i -E "{s/(^ +.*\.pem.*)/# \1/g}" ${$@_FILE}
+	@sed -i -E "{s/(^ +.*\/tls.*)/# \1/g}" ${$@_FILE}
```
I would like to create Secrets with TLS type even when cert-controller is enabled.

Since TLS type Secret needs `tls.crt` and `tls.key` data, we may have to make a  Secret with dummy data (e.g. `dGxzLmNydAo=` ← this is base64 encoded `tls.crt`).

chez-shanpu reviewed (commented):
I left a few comments about what I noticed when I ran e2e locally.

chez-shanpu reviewed (approved):
LGTM!

chez-shanpu reviewed (commented):
I missed one thing.
Could you add flags to `cmd-coil-ipam-controller.md` and `cmd-coil-egress-controller.md` under `docs/`?

p-strusiewiczsurmacki-mobica commented:
> Could you add flags to cmd-coil-ipam-controller.md and cmd-coil-egress-controller.md under docs/?

Done :)

chez-shanpu reviewed (approved):
LGTM👍

chez-shanpu commented:
@p-strusiewiczsurmacki-mobica Thank you for your contribution!
